### ISSUE TITLE
Better documentation for `createLanguage` and `lazy`

### ISSUE THREAD

**Issue by fresheneesz** [jneen/parsimmon#204]
I'd like to create a parameterized parser, much like many of the built in parser combinators (eg `seq` or `string`). I was a little confused as to why I couldn't call the parser functions I defined with `createLanguage` as functions (I found out they're not functions). 

It would be nice to be able to do something like:

```
var L = Parsimmon.createLanguage({
   oneThenAnother: function(a,b) {
       return Parsimmon.seq(a,b)
   },
   something: function() {
       return L.oneThenAnother(string("a"),string("b"))
   }
})
```

But it seems the only parameter that parser functions created by `createLanguage` get is the language namespace that I already have access to via `L`. So it makes me wonder, why am I using `createLangauge` at all? I don't quite understand the importance of the `Lazy` and `createLanguage` functions, which would help clarify this situation: https://github.com/jneen/parsimmon/issues/202.

**Comment by jneen** [comment]
Just make a function?

``` js
function oneThenAnother(a, b) { ... }
```

**Comment by jneen** [comment]
`lazy` and `createLanguage` are to allow forward-declaration of parsers that don't exist yet, so you can have recursive parsers. For example,

``` js
form = seq(lparen, alt(form, atom), rparen)
```
won't work because `form` in the right hand side has value `undefined`. But if you use `lazy`:

``` js
form = lazy(function() { return seq(lparen, alt(form, atom), rparen) })
```

then the `form` on the RHS is not evaluated until the `form` lazy parser has already been constructed. The function you pass to `lazy` is evaluated on the first parse. As far as I understand `createLanguage` is a more convenient way to spread `lazy` across modules.

**Comment by jneen** [comment]
(if you've used haskell's parsec or similar you wouldn't have had to deal with this, because all haskell values are already lazy. unfortunately, js is a strict language, so we have to be explicit about our laziness.)

**Comment by fresheneesz** [comment]
> `form = lazy(function() { return seq(lparen, alt(form, atom), rparen) })`

Wouldn't this work just as well:  `function form() { return seq(lparen, alt(form(), atom), rparen) }`? If that works just as well, I kind of don't understand why `lazy` is even needed, since it just introduces a lot of unnecessary cognitive load.

I assumed there was some optimization trickery going on there. Maybe not?

> we have to be explicit about our laziness

Kind of ironic isn't it ; )

**Comment by jneen** [comment]
Your example would result in a stack overflow if called.

**Comment by fresheneesz** [comment]
Oh right. Well how about `function form() { return seq(lparen, alt(form, atom), rparen) }` then?

**Comment by jneen** [comment]
Then `form` is a function, but `alt` expects a parser.

**Comment by wavebeem** [comment]
Hmm, I understand your desire to put a function in there, but it's not really compatible with the point of `createLanguage` which is just to delay evaluation of parsers so you don't have to manually use `lazy` ever. 

Your function will need to be outside of `createLanguage`, like @jneen said. Maybe I'll add this as advice to the `createLanguage` documentation.

**Comment by wavebeem** [comment]
I think maybe I need to cover a little bit more of the "why" on the `createLanguage` documentation, and add "don't use this for parameterized parsers"

**Comment by fresheneesz** [comment]
Mk, I gotcha. I think what I got from the documentation is "you should be using `createLangauge`" and so I said "ok its probably important". But had I understood the system, I'd have started by creating normal functions and using `lazy` when/if necessary. 

I'm trying to think of how to improve the API for this, so sorry if I'm being a bit obtuse.

**Comment by wavebeem** [comment]
No problem. I would like to be able to integrate parameterized parsers with
it but couldn't figure out a good way. Let me know if you can figure it
out. It might have to be a new function or breaking change though.

**Comment by fresheneesz** [comment]
@wavebeem Hmm, best idea I can think of is returning a function from calling `Parsimmon` that you place all the `Parsimmon` instance properties (including methods) onto. That way you can use it either raw or by calling the function. Eg:

```
function Parsimmon(action) {
  var instance = function() { 
    action.apply(instance, arguments)
  } 
  instance._ = action;
  for(var method in parsimmonMethods) {
    instance[method] = parsimmonMethods[method]
  }  

  return instance
}
```

So if your parser requires arguments, you call it with its arguments to return a combinator (eg `seq(str('.'), myParser(whatever))`, and if it doesn't have arguments, you just use it raw (eg `seq(str('.'),myParser)`. Would that work?

**Comment by wavebeem** [comment]
That would require creating a custom parser and wouldn't solve the lazy problem. I could change `createLanguage`  to make an object with METHODS that return parsers, rather than just parsers (I like this idea---it would catch errors even sooner because of missing method calls), but it would be a breaking change to the API, so it will need to wait till Parsimmon 2.0.

**Comment by wavebeem** [comment]
Here's an example of what I mean:

```js
const P = require("parsimmon");

P.createLanguage = parsers => {
  const lang = {};
  for (const [key, val] of Object.entries(parsers)) {
    lang[key] = (...args) => P.lazy(() => val.apply(lang, args));
  }
  return lang;
};

const Lisp = P.createLanguage({
  Expr() {
    return P.alt(
      this.Number(),
      this.List()
    );
  },

  List() {
    return P.alt(
      this._list("(", ")"),
      this._list("[", "]"),
      this._list("{", "}")
    )
  },

  _list(before, after) {
    const n = arguments.length;
    if (n !== 2) {
      throw new Error(`bad argument count: ${n}`);
    }
    return this.Expr()
      .trim(this._())
      .many()
      .wrap(P.string(before), P.string(after));
  },

  _() {
    return P.optWhitespace;
  },

  Number() {
    return P.regexp(/[0-9]+/);
  },
})

const val = Lisp.Expr().tryParse("(1 2)");
console.log(val);
```

**Comment by jneen** [comment]
Is using `Function.prototype.length` considered bad form these days

**Comment by wavebeem** [comment]
I generally avoid it except to write argument count checks. Not that there's anything particularly wrong about it, just feels a little magical to change behavior based on it since JS is so lax about args to begin with.

**Comment by wavebeem** [comment]
I'm assuming you bring this up because I could check if the function takes >1 arguments and make it a function that returns a parser rather than just a parser? Technically I could do it, though it is inconsistent.

Honestly I've been wondering about a new replacement for `createLanguage` in general after reading up about stateful parser combinators in Scala and coming up with https://github.com/jneen/parsimmon/blob/master/examples/python-ish.js to manage state.

Something that explicitly had a `state` property (probably defaults to `undefined`) and a `.update(...)` which returned a new language with a new state (no mutation). The rest of it could work like the modified `createLanguage` I just posted.

**Comment by fresheneesz** [comment]
> Is using Function.prototype.length considered bad form these days

It breaks when you use `arguments` instead of listing the arguments normally.

**Comment by fresheneesz** [comment]
> an object with METHODS that return parsers

That works but then you always have to call the functions. I'm trying to remove as much boilerplate as possible. 

Check out the last two commits on my experimental branch. I have something like the following working: 

```
var L = createLanguage({
    a: "a",
    b: /b/,
    c: string("c"),
    d: function() {
        return string('d')
    },
    e: function(e) {
        return string(e)
    },
    recursive: function() {
        return seq('f', recursive.atMost(1))
    },
    start: function() {
        return seq(a,b,c,d,e('e'),recursive)
    }
})

// ...

console.log(start.tryParse('abcdefff'))
```

https://github.com/fresheneesz/parsimmon/commits/experimental

**Comment by fresheneesz** [comment]
Oops, I didn't do the recursion test right, and it does break things. One sec.

**Comment by fresheneesz** [comment]
Ok, I fixed up my branch with my last commit so that recursion works and all the rest. I updated my above comment to fix the example there too. So check the last *3* commits in my experimental branch.

**Comment by wavebeem** [comment]
> That works but then you always have to call the functions. I'm trying to remove as much boilerplate as possible.

Yeah but it means you catch errors sooner and easier, because it tells you what method was missing.

Also, recognizing `"f"` as a parser instead of `string("f")` is not a goal I have, and there's explicitly work to remove that kind of behavior in Parsimmon.

**Comment by fresheneesz** [comment]
> it means you catch errors sooner and easier

Hmm, that might make it a good tradeoff.

> recognizing "f" as a parser instead of string("f") is not a goal I have, and there's explicitly work to remove that kind of behavior in Parsimmon

Why remove that kind of behavior? 

Also, what are your thoughts on the way I enabled createLanguage to support parameterized parsers? Seems like a backwards compatible change. If you don't want the string/regex -> parser mapping, that can be removed.

**Comment by wavebeem** [comment]
I think it keeps it clearer that parsers are what get combined into new parsers, which is key to the power of the library.

Also I've personally found this kind of behavior confusing in Lodash.

**Comment by wavebeem** [comment]
Also you may want to know that rather than using eval you can use regular variables and `lazy` like this:

```js
var Expr = lazy(() => P.alt(A, B));

var A = ...;

var B = ...;
```

And you just need to write `lazy(() => ...)` around parsers that reference parsers defined below it.

**Comment by fresheneesz** [comment]
> I've personally found this kind of behavior confusing in Lodash

Hmm, I'm curious what an example would be that is confusing in lodash. 

> rather than using eval

I'm also using eval to splay out core Parsimmon combinators, to eliminate boilerplate. 

Still looking for feedback on my createLangauge parameterized parsers.

**Comment by wavebeem** [comment]
> lodash

For example, `_.filter` accepts strings, numbers, objects, and functions as parameters. This can make sense in the case of literal calls, but it also makes it very easy to accidentally pass the wrong parameter and get no errors but very confusing results.

> eval

I mean you do you, but I find it very confusing and it made it a bit confusing to follow your examples for me.

> parameterized parsers

Can you summarize the change? There were several commits and lots and lots of things mixed together in there and I don't have a ton of time to scan through it right now.

**Comment by fresheneesz** [comment]
> makes it very easy to accidentally pass the wrong parameter and get no errors but very confusing results

Looking at the docs for `lodash.filter`, they have pretty bad documentation. The arguments doesn't even explain that the "Function" can actually be a string, or array, or object, or what passing something htat's not a function there would even mean. Their example is pretty obtuse to someone like me that isn't well versed in lodash. 

Would allowing strings, regex, and functions cause the same problem in parsimmon? I don't think so. A combination of the fact that parsimmon has a far simpler API (even with string and regex promotion) and that its documentation doesn't have to suck like lodash's means I don't think we'll have that kind of problem.

> Can you summarize the change?

In [this commit](https://github.com/fresheneesz/parsimmon/commit/df9dfa19691e5d4e3ab09eaefcf91632796efcb9), I used the approach I suggested before to return a callable function that has the `_` action set on it, and when called, passes through the call to that same `_` action, so that `combinator()` does the same thing as `combinator._()`. There's also some extraneous stuff related to mapping strings/regex to parsers that you can ignore.

In [this commit](https://github.com/fresheneesz/parsimmon/commit/256707097436e4748ef07db7d2852f13bec7b4d4), I updated `lazy` so that its return value also can be called either like `f()` or like `f._()`.

**Comment by wavebeem** [comment]
Well, as of right now you're depending on a nonstandard JS feature (`__proto__`), and also relying on an ES6 function (`Object.assign`). Also it appears to be mutating the Parsimmon.prototype on accident, maybe?

Parsimmon currently supports IE7, so the change definitely couldn't come in as-is.

I'm feeling pretty set about using `this` as a reference to the other rules, as well making all rules use functions because it provides better error protection against typos and keeps things consistent.

Plus I want to think about how to easily support stateful parsers directly within Parsimmon, so I kinda wanna do a bit of an overhaul or replacement for `createLanguage` with that in mind.

**Comment by fresheneesz** [comment]
> nonstandard JS feature (__proto__)

That's why my original earlier in this thread just copied the prototype methods. Easy to replace `Object.assign` too.

> I'm feeling pretty set about using this as a reference to the other rules

Mk, I'm curious to see where that line of thinking goes. I definitely think better error handling is more important than removing boilerplate.

**Comment by wavebeem** [comment]
I should probably have some time tonight to make a branch with an example of what I mean for a (optionally) stateful and consistent replacement for `createLanguage`.

**Comment by wavebeem** [comment]
https://runkit.com/wavebeem/59cdac70837de0001248bd64

(If you've never used RunKit (formerly known as Tonic), click the little refresh icon at the bottom to see the code output.)

That's my idea for the new API right now. The only major problem I see is that you can't have helpers with no arguments that return objects other than parsers.

I think it's a reasonable tradeoff that functions with arguments don't get wrapped in `lazy` for you, and they probably wouldn't need to be? I'm having a hard time thinking of a case where that'd be necessary.

**Comment by fresheneesz** [comment]
Cool. How would the `state` be used? Would be nice to see an example of that - or are you already using state and I'm just not seeing it?

> it's a reasonable tradeoff that functions with arguments don't get wrapped in `lazy` for you

Couldn't we combine this with the technique I was using so that lazy is still called for you?

**Comment by wavebeem** [comment]
```js
  Indent() {
    return this.countWhitespace().chain(indent => {
      const n = this.state[this.state.length - 1];
      if (indent !== n) {
        return P.fail("equal indentation");
      }
      return P.of(n);
    });
  },
```

I'm using `this.state` right there to keep track of the current indentation level and assert indentation gets deeper as blocks are nested.

Also it is calling `lazy` already:

```js
const parser = P.lazy(func);
```

**Comment by fresheneesz** [comment]
Ok ok dunno how I missed that. I'm still not quite sure how that's working (or what its doing). I don't see anything actually setting or modifying any properties on the state. I just see the state being initialized to `[0]` and don't see anywhere that would change.

**Comment by wavebeem** [comment]
The important part is `withState` which returns a `new Language(...)` that has the new state on it, then we use the parsers attached to the new state object.

**Comment by fresheneesz** [comment]
Ok, I see what you're doing now. I'm still not used to ES6. Is that a splat operator I see? So you're essentially creating a new state with the indentation appended onto the end of the list. `[...this.state, indent]` is the same as `this.state.concat([indent])` right? 

So does this mean you need to use `withState` on every parser you want to pass any state to? I guess that's not so bad. It would definitely reduce the need for a second pass in my parser.

**Comment by wavebeem** [comment]
> [...this.state, indent] is the same as this.state.concat([indent]) right?

Yeah.

> So does this mean you need to use withState on every parser you want to pass any state to? I guess that's not so bad. It would definitely reduce the need for a second pass in my parser.

I might change the name of that method, but yeah that's the idea.

**Comment by fresheneesz** [comment]
> withState

So actually, is that really much better than just passing state into parser functions? Like, you're only really removing the need to add a `state` parameter to all your stateful parsers (and adding the need to write `this.` so what are you really saving?). 

What I'd suggest is that you keep `withState` in order to reset state, but for everything else, pass that state through automatically. That way you get state without the boilerplate. 

The one thing that increases complexity here tho is that people have to still be careful about changing state - ie if an inner parser succeeds and adds a state property (or modifies a state property) but its caller fails, that state could/would remain modified. You could mitigate that by always making copies of the state before changing it (eg by making `state` a getter/setter situation. Do you think that's a reasonable thing to do?

Also, since combinators themselves never have any persistent state, you should always be able to use `Language.combintaor` in place of `this.combinator`, right? Just want to make sure, cause this could still allow me to remove all the `this.` boilerplate that your proposal suggests.

**Comment by wavebeem** [comment]
The other API I considered is:

```js
const Lang = P.createLanguage([0], {
  Indent(parsers, state) {
    return parsers.Indent(state); // something like that, no `this`
  }
});
```

Where state and parsers are passed in to each function. It's a bit more explicit that way.

But the whole point of keeping state on `this.state` is that everything automatically has access to state and you don't have to manually pass it around. With my approach you can give the language an initial state, and you can make new sub-languages with new states, as opposed to manually passing the state around.

Parsimmon is already incompatible with state mutations since there's no cleanup step if a parse fails later on, so you would have to avoid mutation with `createLanguage` too. So you'd have to use `state.concat(...)` and `Object.assign({}, state, {...})` (or ES6 `[...state, blah]` and `{...state, blah}`).

> Also, since combinators themselves never have any persistent state, you should always be able to use Language.combintaor in place of this.combinator, right? Just want to make sure, cause this could still allow me to remove all the this. boilerplate that your proposal suggests.

I would hardly call `this` boilerplate in the example. Parsers and state need to be scoped to the particular language being created, so they have to be properties or methods on *something*. I suppose you could use `Language.combinator` instead and that *would* work, but it sorta defeats the entire point of being an object method and getting `this` for free now that I think about it. And it avoids having to repeat the language name everywhere. That's why `parsers` is passed as an argument in the current version of `createLanguage`.

**Comment by fresheneesz** [comment]
> as opposed to manually passing the state around

But aren't you basically manually passing the state around with `withState` anyway, in your proposal? 

> they have to be properties or methods on something

Ok I think you're right here.

**Comment by wavebeem** [comment]
> But aren't you basically manually passing the state around with withState anyway, in your proposal?

It's all managed internally by the Language class inside `createLanguage` though. You never have to do something like `this.someParser(state)`. You would have to pass `state` as the first parameter to *every* function in case anywhere down the chain needed it. That's literally the whole point of classes and `this`, to automatically thread a piece of state around.

Maybe I'm misunderstanding what you're asking for? If you could post an example that might be helpful. But it just sounds harder the way I'm understanding your suggestion.

**Comment by wavebeem** [comment]
My guess is that you mean something like this:

```js
const Python = P.createLanguage({
  Block(state) {
    return P.string("block:\n")
      .then(this.countWhitespace(state))
      .chain(indent => {
        const n = this.state[this.state.length - 1];
        if (indent <= n) {
          return P.fail("deeper indentation");
        }
        return P.seqMap(
          this.Statement(state),s
            .IndentedStatement([...this.state, indent])
            .many(),
          (first, rest) => ["Block", [first, ...rest]]
        );
      });
  },

  Indent(state) {
    return this.countWhitespace(state).chain(indent => {
      const n = this.state[this.state.length - 1];
      if (indent !== n) {
        return P.fail("equal indentation");
      }
      return P.of(n);
    });
  },

  IndentedStatement() {
    return this.Indent(state).then(this.Statement(state));
  },

  Call() {
    return P.regexp(/[a-z]+/i)
      .skip(P.string("()\n"))
      .map(s => ["Call", s, []]);
  },

  Statement() {
    return P.alt(this.Call(state), this.Block(state));
  },

  countWhitespace() {
    return P.optWhitespace.map(s => s.length);
  }
});
```

Which is awkward to pass `state` everywhere, and I don't know really know how to implement `createLanguage` such that that works off the top of my head.

**Comment by fresheneesz** [comment]
> awkward to pass state everywhere

I wasn't actually suggesting that - I was comparing `withState` to that. Here's my line of thinking. I'd like to write something like:

```
something: function() {
  return this.whitespace().chain
       return withState({line: this.line, this.column}).seq(
          this.statement(),
          this.whitespace()
       )
  )
}
```

But will `this.statement` and `this.whitespace` get the given state i passed to `withState`? If not, then I'd have to use withState for each item in the sequence right?

**Comment by wavebeem** [comment]
Ah, I see. In that case you'd need something like this:

```js
something: function() {
  return this.whitespace().chain(() => {
       const x = withState({line: this.line, this.column})
        return P.seq(
          x.statement(),
          x.whitespace()
       )
 })
}
```

Since `withState` returns a new `Language` you would have to chain off of the new object.

**Comment by wavebeem** [comment]
also for what it's worth, this is all basically an evolution of my idea from https://github.com/jneen/parsimmon/blob/master/examples/python-ish.js#L20

**Comment by fresheneesz** [comment]
That alternate way of doing `withState` looks like it'd work pretty well

**Comment by wavebeem** [comment]
This has gotten off topic and we can move discussion to #209.
